connector_office_365_hr_leave/models/hr_leave.py

```diff
@@ -54,7 +68,31 @@ def _validate_leave_request(self):
 
     @api.multi
     def action_refuse(self):
-        return super(
-            HolidaysRequest,
-            self.with_context(o365_override_user=True)
-        ).action_refuse()
+        res = False
+        activity_scheduled = set()
+        now = fields.Datetime.now()
+        leaves_non_expired_tokens = self.filtered(
+            lambda r: r.meeting_id.user.office_365_access_token and
+            r.meeting_id.office_365_expiration > now
+        )
+        leaves_expired_tokens = self - leaves_non_expired_tokens
+        for rec in leaves_non_expired_tokens:
+            try:
+                res = super(
+                    HolidaysRequest,
+                    rec.with_context(o365_override_user=True)
+                ).action_refuse()
+            except Office365Error:
+                leaves_expired_tokens += rec
+        for rec in leaves_expired_tokens:
+            # don't delete event from office365, create an activity
+            # requesting the user to auth on office (at most 1 activity per user)
+            res = super(
+                HolidaysRequest,
+                rec.with_context(office_365_force=True)
+            ).action_refuse()
+            # don't flood the user with activities
+            if rec.user_id not in activity_scheduled:
+                self._office_365_activity_require_auth()
```

Comment:
```suggestion
                rec._office_365_activity_require_auth()
```
self is recordset here, no?

Reply:
yes you are right